### PR TITLE
Inventories: Collector

### PR DESCRIPTION
**Pull request by albertvaka**
### What does this PR do?

- Adds a collector for inventories, that can be enabled with the config option `inventories_enabled`.
- Tweaks the SendNow function in Scheduler to add a delay parameter
- Adds a test for the collector

### Additional notes
I'm not adding a changelog note since this feature can't still be used by our users.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -548,6 +548,8 @@ func initConfig(config Config) {
 	config.SetKnown("apm_config.receiver_timeout")
 	config.SetKnown("apm_config.watchdog_check_delay")
 
+	config.BindEnvAndSetDefault("inventories_enabled", false)
```
Comment (hush-hush) [inline pkg/config/config.go]: Should we not enable it by default ?

File: cmd/agent/app/run.go [hunk 2]
```diff
@@ -284,6 +284,12 @@ func StartAgent() error {
 		return err
 	}
 
+	if config.Datadog.GetBool("inventories_enabled") {
```
Comment (hush-hush) [inline cmd/agent/app/run.go]: This needs to be done in the dogstatsd binary too.

File: pkg/metadata/inventories/inventories.go [hunk 3]
```diff
@@ -61,8 +82,25 @@ func SetCheckMetadata(checkID, key string, value interface{}) {
...
+		entry.CheckInstanceMetadata[key] = value
+
+		select {
+		case metadataUpdatedC <- nil:
+		default: // To make sure this call is not blocking
+		}
+	}
+}
+
+// Only used in tests
```
Comment (hush-hush) [inline pkg/metadata/inventories/inventories.go]: I think we could move this function to the test files then.

File: pkg/metadata/inventories/inventories.go [hunk 4]
```diff
@@ -78,15 +116,17 @@ func getCheckInstanceMetadata(checkID, configProvider string) *CheckInstanceMeta
...
 	checkInstanceMetadata["config.hash"] = checkID
 	checkInstanceMetadata["config.provider"] = configProvider
 
 	return &checkInstanceMetadata
 }
 
 // GetPayload fills and returns the check metadata payload
-func GetPayload(ac getLoadedConfigsInterface, coll getAllInstanceIDsInterface) *Payload {
+func GetPayload(hostname string, ac autoConfigInterface, coll collectorInterface) *Payload {
+	lastPayloadSent = timeNow()
```
Comment (hush-hush) [inline pkg/metadata/inventories/inventories.go]: Could we rename this to `lastPayloadCreated` ?

Comment (hush-hush) [inline pkg/metadata/inventories/inventories.go]: Should we use the `Atomic`package here to write `lastPayloadSent` and line 170 to read it?

File: pkg/metadata/scheduler.go [hunk 5]
```diff
@@ -101,13 +105,25 @@ func (c *Scheduler) AddCollector(name string, interval time.Duration) error {
...
+	_, found := c.collectors[name]
+	return found
+}
+
+// SendNow runs a collector manually, on demand. The delay parameter can be set
+// to 0 to run it immediately, or to a duration after which the collector will
+// run. The runs at regular intervals of the collector will be resumed *after*
+// this manual run. Calling SendNow on a stopped Scheduler does nothing, since
+// the goroutine that waits on the Timer will not be running.
+func (c *Scheduler) SendNow(name string, delay time.Duration) {
```
Comment (hush-hush) [inline pkg/metadata/scheduler.go]: Could we rename this, I don't think the name describes the function clearly. `TriggerAndResetCollectorTimer` maybe ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
My main issue with the "send as soon as we have a change" approach is that we're always going to send an almost empty payload at first (most likely containing only the `hostname_source`) and then wait 5 minutes to sent the first complete/usable payload.

If we send the  payload every 5 minutes I don't think we need to send it sooner on an update. Do we really need this (for 6.15) ? I don't think current behavior is better than just an short heartbeat.

File: pkg/config/config.go (hunk 1 above)
Comment (albertvaka) [reply pkg/config/config.go]: It's a beta thing, so I assumed not.

File: pkg/metadata/scheduler.go (hunk 5 above)
Comment (albertvaka) [reply pkg/metadata/scheduler.go]: It was hard for me to come up with a good name. `TriggerAndResetCollectorTimer` is pretty descriptive, I'll change it.

**Comment by albertvaka** [comment]
> My main issue with the "send as soon as we have a change" approach is that we're always going to send an almost empty payload at first (most likely containing only the `hostname_source`) and then wait 5 minutes to sent the first complete/usable payload.

The first packet shouldn't be sent after 5 minutes after the agent is running (because lastPayloadSent is initialized to current time, so it's queued after `minSendInterval`).

> 
> If we send the payload every 5 minutes I don't think we need to send it sooner on an update. Do we really need this (for 6.15) ? I don't think current behavior is better than just an short heartbeat.

It was a requirement when we designed this to have closer to real-time information, but probably not a strong one that we could drop if the code becomes too complicated. My main argument for keeping it is: The min interval of 5 minutes is quite conservative now, but if we change it to 1 minute it can make a difference from a product point of view.

File: pkg/metadata/inventories/inventories.go (hunk 4 above)
Comment (albertvaka) [reply pkg/metadata/inventories/inventories.go]: I think `Atomic` only has support for basic times and this is a `time.Time` :/

Comment (hush-hush) [reply pkg/metadata/inventories/inventories.go]: Do we need a mutex then ?

File: pkg/config/config.go (hunk 1 above)
Comment (hush-hush) [reply pkg/config/config.go]: @ItsYamnitsky What should be the default ?

File: pkg/metadata/inventories/inventories.go (hunk 4 above)
Comment (hush-hush) [reply pkg/metadata/inventories/inventories.go]: We should add a mutex then, no ?

Comment (albertvaka) [reply pkg/metadata/inventories/inventories.go]: Added.

**Review by hush-hush — APPROVED** [review]

**Pull request merged**